File: hack/integration/faults.go
package main

import (
	"fmt"
	"strings"
	"time"
)

// exerciseFaults disrupts one PersistentFleet member at a time in each way a
// kind cluster can (ADR 0024 qualification): graceful and forced Pod
// deletion, SIGKILL of celld, a node drain serialized by the PDB, manager
// kills mid-rollout and mid-contraction, a deleted claim, a lost volume and a
// node whose kubelet stops. It then takes every member down at once, by
// SIGKILL and by Pod deletion. Continuous writes run through every fault, and
// after each one the fleet must settle with every acknowledged write readable.
// No step is repaired by hand.
func (h *harness) exerciseFaults() {
	h.scale("beta", 3)
	h.writeLedger("beta")

	h.fault("graceful Pod delete", func() {
		h.k("-n", "fleets", "delete", "pod", "beta-1", "--wait=false")
	}, h.sameDisksAfter(), h.newPod("beta-1"))

	h.fault("forced Pod delete", func() {
		h.k("-n", "fleets", "delete", "pod", "beta-0", "--force", "--grace-period=0", "--wait=false")
	}, h.sameDisksAfter(), h.newPod("beta-0"))

	h.fault("SIGKILL of celld", func() {
		h.sigkill("beta-2")
		h.wait("kubelet restarts the killed celld container", func() bool {
			h.observe(h.faultWatch)
			return h.restartCounts()["beta-2"] > h.faultBefore.restarts["beta-2"]
		})
	}, h.sameDisksAfter(), h.restarted("beta-2"))

	h.fault("node drain serialized by the PDB", h.drainTwoNodes, h.sameDisksAfter(), h.newPod("beta-1"))

	h.fault("manager killed mid-contraction", func() {
		h.shrinkWatching(2, nil, h.killOperator)
		h.setReplicas("beta", 3)
	}, h.sameDisksAfter())

	// The StatefulSet controller recreates a deleted claim for the member's
	// next Pod; the operator takes no part.
	h.fault("deleted claim", func() {
		h.k("-n", "fleets", "delete", "pvc", "data-beta-1", "--wait=false")
		h.k("-n", "fleets", "delete", "pod", "beta-1", "--wait=false")
		h.waitReplacedDisk("data-beta-1", 10*time.Minute)
	}, h.freshDiskAfter("data-beta-1"))

	h.fault("lost volume", func() {
		// A backend disk that no longer exists: the PV object is removed
		// without its CSI and protection finalizers, as after an EBS loss.
		// Kubernetes marks the claim Lost and the operator replaces the member.
		pv := str(h.get("pvc", "data-beta-0"), "spec", "volumeName")
		h.k("delete", "pv", pv, "--wait=false")
		h.k("patch", "pv", pv, "--type=merge", "-p", `{"metadata":{"finalizers":null}}`)
		h.waitReplacedDisk("data-beta-0", 10*time.Minute)
	}, h.freshDiskAfter("data-beta-0"))

	victim, node := h.memberOnSpareNode("beta")
	h.fault("node failure", func() { h.failNode(node, victim) }, h.freshDiskAfter("data-"+victim))

	h.fault("manager killed mid-rollout", func() {
		h.rollAll("beta", encode(object{"spec": object{"maintenance": object{"restartToken": "faults-rollout"}}}), nil, h.killOperator)
	})

	// Every member at once: celld restarts all of them on their own disks and
	// they recover each other's logs, with no action from anyone.
	h.fleetOutage("every celld killed at once", func() {
		for _, pod := range h.memberPods("beta") {
			h.sigkill(nameOf(pod))
		}
		h.wait("kubelet restarts every killed celld container", func() bool {
			after := h.restartCounts()
			for name, before := range h.faultBefore.restarts {
				if after[name] <= before {
					return false
				}
			}
			return true
		})
	}, h.restartedInPlace())
	h.fleetOutage("every Pod deleted at once", func() {
		h.k("-n", "fleets", "delete", "pod", "-l", "celld.eric.dev/fleet-uid="+uidOf(h.get("celldfleet", "beta")), "--wait=false")
		h.wait("the StatefulSet replaces every Pod", func() bool {
			after := nameUIDs(h.memberPods("beta"))
			for name, uid := range h.faultBefore.pods {
				if after[name] == "" || after[name] == uid {
					return false
				}
			}
			return true
		})
	})

	// A Bucket member is disposable: its disk is a cache.
	h.scale("alpha", 2)
	h.writeLedger("alpha")
	h.startWriter("alpha")
	h.k("-n", "fleets", "delete", "pod", "alpha-1", "--force", "--grace-period=0", "--wait=false")
	h.waitSettled("alpha", 2, 5*time.Minute)
	h.stopWriter("alpha")
	h.readLedger("alpha")
	fmt.Println("PASS: every single-member fault and whole-fleet outage was absorbed automatically with every acknowledged write readable")
}

// fault runs inject against beta under continuous writes, waits for beta to
// settle at three members while watching the one-disruption invariant, runs
// the checks, and verifies the ledger.
func (h *harness) fault(name string, inject func(), checks ...func()) {
	fmt.Println("FAULT:", name)
	h.waitSettled("beta", 3, 10*time.Minute)
	h.startWriter("beta")
	before := h.claims("beta")
	pods := nameUIDs(h.memberPods("beta"))
	h.faultBefore = faultSnapshot{claims: before, pods: pods, restarts: h.restartCounts()}
	d := h.watchDisruptions("beta")
	h.faultWatch = d
	inject()
	h.waitWatching("beta settles after "+name, 15*time.Minute, d, func() bool { return h.settled("beta", 3) })
	for _, check := range checks {
		check()
	}
	h.stopWriter("beta")
	h.readLedger("beta")
	fmt.Println("PASS: fault absorbed:", name)
}

// fleetOutage takes down every beta member at once under continuous writes.
// The one-disruption invariant does not apply; the fleet must settle again on
// its own disks with every acknowledged write readable and no disk replaced.
func (h *harness) fleetOutage(name string, inject func(), checks ...func()) {
	fmt.Println("FAULT:", name)
	h.waitSettled("beta", 3, 10*time.Minute)
	h.startWriter("beta")
	h.faultBefore = faultSnapshot{claims: h.claims("beta"), pods: nameUIDs(h.memberPods("beta")), restarts: h.restartCounts()}
	inject()
	h.waitFor("beta recovers from "+name, 15*time.Minute, func() bool { return h.settled("beta", 3) })
	must(sameDisks(h.faultBefore.claims, h.claims("beta")))
	for _, check := range checks {
		check()
	}
	h.stopWriter("beta")
	h.readLedger("beta")
	fmt.Println("PASS: whole-fleet outage absorbed:", name)
}

// restartedInPlace requires every member to keep its Pod: kubelet restarted
// the containers and nothing replaced a member.
func (h *harness) restartedInPlace() func() {
	return func() {
		assert(same(nameUIDs(h.memberPods("beta")), h.faultBefore.pods), "a member Pod was replaced after its container was killed")
	}
}

type faultSnapshot struct {
	claims   map[string]claimIdentity
	pods     map[string]string
	restarts map[string]int64
}

func (h *harness) sameDisksAfter() func() {
	return func() { must(sameDisks(h.faultBefore.claims, h.claims("beta"))) }
}

func (h *harness) newPod(name string) func() {
	return func() {
		assert(nameUIDs(h.memberPods("beta"))[name] != h.faultBefore.pods[name], "%s kept its Pod", name)
	}
}

func (h *harness) restarted(name string) func() {
	return func() {
		assert(nameUIDs(h.memberPods("beta"))[name] == h.faultBefore.pods[name], "%s was replaced instead of restarted", name)
		assert(h.restartCounts()[name] > h.faultBefore.restarts[name], "%s container did not restart", name)
	}
}

// freshDiskAfter requires the member's disk replaced and every other disk kept.
func (h *harness) freshDiskAfter(claim string) func() {
	return func() {
		after := h.claims("beta")
		assert(freshDisk(h.faultBefore.claims[claim], after[claim]), "%s was not replaced by a new disk: %+v", claim, after[claim])
		for name, old := range h.faultBefore.claims {
			if name != claim {
				assert(after[name] == old, "replacing %s changed %s", claim, name)
			}
		}
	}
}

// waitReplacedDisk waits for a new claim of the same name after a claim was
// deleted. The fleet has not necessarily noticed the loss when inject
// returns, so settling alone would prove nothing.
func (h *harness) waitReplacedDisk(claim string, timeout time.Duration) {
	old := h.faultBefore.claims[claim].UID
	h.waitFor("replacement of "+claim, timeout, func() bool {
		h.observe(h.faultWatch)
		current, ok := h.claims("beta")[claim]
		return ok && current.UID != old
	})
}

// restartCounts reports the celld container restart count of each beta member.
func (h *harness) restartCounts() map[string]int64 {
	out := map[string]int64{}
	for _, pod := range h.memberPods("beta") {
		for _, c := range list(pod, "status", "containerStatuses") {
			if str(c, "name") == "celld" {
				out[nameOf(pod)] = num(c, "restartCount")
			}
		}
	}
	return out
}

// sigkill kills celld's process from its node, bypassing SIGTERM's drain.
func (h *harness) sigkill(podName string) {
	pod := h.get("pod", podName)
	node := str(pod, "spec", "nodeName")
	var id string
	for _, c := range list(pod, "status", "containerStatuses") {
		if str(c, "name") == "celld" {
			_, id, _ = strings.Cut(str(c, "containerID"), "://")
		}
	}
	assert(node != "" && id != "", "cannot locate celld container of %s", podName)
	pid := strings.TrimSpace(h.run(command{args: []string{"docker", "exec", node, "crictl", "inspect", "--output", "go-template", "--template", "{{.info.pid}}", id}, timeout: 30 * time.Second}))
	h.sh(30*time.Second, "docker", "exec", node, "kill", "-9", pid)
}

// drainTwoNodes drains beta-1's node, which the budget allows, and then
// requires a second drain to be refused while beta-1 cannot return: its disk
// is on the cordoned node, so beta-1 stays unready and the fixed budget of one
// admits no further eviction.
func (h *harness) drainTwoNodes() {
	selector := "--pod-selector=celld.eric.dev/fleet-uid=" + uidOf(h.get("celldfleet", "beta"))
	first := str(h.get("pod", "beta-1"), "spec", "nodeName")
	second := str(h.get("pod", "beta-0"), "spec", "nodeName")
	survivor := uidOf(h.get("pod", "beta-0"))
	defer func() {
		h.k("uncordon", first)
		h.k("uncordon", second)
	}()
	h.k("drain", first, selector, "--ignore-daemonsets", "--delete-emptydir-data", "--timeout=180s")
	h.wait("PDB allows no disruption while beta-1 is down", func() bool { return h.budget("beta") == 1 && h.disruptionsAllowed("beta") == 0 })
	out, err := h.tryK("drain", second, selector, "--ignore-daemonsets", "--delete-emptydir-data", "--timeout=30s")
	assert(err != nil, "second drain evicted a member while the fleet was recovering: %s", out)
	assert(uidOf(h.get("pod", "beta-0")) == survivor, "second drain replaced beta-0")
	fmt.Println("PASS: PDB refused a second eviction while a drained member was down")
}

// memberOnSpareNode picks a member on a node that neither the store nor the
// operator runs on, so failing that node disrupts only the fleet.
func (h *harness) memberOnSpareNode(fleetName string) (member, node string) {
	store := str(h.getIn(storeNS, "pod", "minio"), "spec", "nodeName")
	for _, pod := range h.memberPods(fleetName) {
		if n := str(pod, "spec", "nodeName"); n != "" && n != h.nodes[0] && n != store {
			return nameOf(pod), n
		}
	}
	fail("no %s member runs on a node free of the store and the operator", fleetName)
	return "", ""
}

// failNode stops a node's kubelet, as when a node hangs or loses the control
// plane. Its containers keep running, so celld on it becomes a zombie. After
// the unreachable-node toleration the member's Pod is evicted and stays
// Terminating, since nothing confirms it stopped. The operator must
// force-delete it and, once the member has been down for the replacement
// delay, replace its disk. The kubelet returns only after both, and the
// member rejoins on a fresh disk.
func (h *harness) failNode(node, member string) {
	old := h.faultBefore.pods[member]
	h.sh(time.Minute, "docker", "exec", node, "systemctl", "stop", "kubelet")
	defer h.sh(time.Minute, "docker", "exec", node, "systemctl", "start", "kubelet")
	h.waitFor("operator force-deletes "+member+" from the failed node", 15*time.Minute, func() bool {
		h.observe(h.faultWatch)
		pod, err := h.tryGet("fleets", "pod", member)
		return err != nil || uidOf(pod) != old
	})
	h.waitReplacedDisk("data-"+member, memberReplacementDelay+10*time.Minute)
}

File: hack/integration/faults.py
"""Fault injection in the harness-owned kind cluster.

Call only from run.py --faults after its identity, isolation and drift fixtures
pass. Scenarios, in order:

1. Manager crash after the replica CAS, before the journal records it (scale-out).
2. Manager crash after durable intent, before the replica CAS (contraction).
3. Node loss for a Bucket replica: cordon plus pod deletion, contraction blocked.
4. Node loss for a launcher-managed PersistentFleet replica: retained claim,
   no decrement, same-host return.
5. S3 latency through toxiproxy: contraction still completes, write readable.
6. S3 partition through toxiproxy while a contraction is issued: no further
   decrement, no completion or loss from absent evidence, write readable after
   recovery, and whatever the post-outage outcome is, exactly one effect.

Every assertion is about fail-closed behavior. This module never discovers a
kubeconfig, touches AWS, or changes anything outside the harness cluster.
"""
import json
import time

JOURNAL = 'celld.example.com/lifecycle-journal'
OPERATION = 'celld.example.com/lifecycle-operation'


def exercise(k, apply, get, wait_for, ready, curl, probe, cluster, curl_image, set_operator_fault):
    def journal(fleet):
        values = json.loads(k('get', 'celldstoragereservations', '-o', 'json'))['items']
        value = next(v for v in values if v['spec']['fleetName'] == fleet and v['spec']['fleetNamespace'] == 'fleets')
        return json.loads(value['metadata'].get('annotations', {}).get(JOURNAL, 'null'))

    def workload(fleet):
        kind = 'deployment' if get('celldfleet', fleet)['spec']['profile'] == 'Bucket' else 'statefulset'
        return get(kind, fleet)

    def ready_reason(fleet):
        for c in get('celldfleet', fleet).get('status', {}).get('conditions', []):
            if c['type'] == 'Ready':
                return c['reason']
        return ''

    def operation_id(fleet):
        return get('celldfleet', fleet).get('status', {}).get('lifecycle', {}).get('operationID')

    def settled(fleet, count):
        return workload(fleet)['spec']['replicas'] == count and ready(fleet) and not journal(fleet).get('Operation')

    def completions(fleet, op):
        return [h for h in journal(fleet).get('History', []) if h['ID'] == op]

    def fleet_pods(fleet):
        uid = get('celldfleet', fleet)['metadata']['uid']
        return json.loads(k('-n', 'fleets', 'get', 'pods', '-l', 'celld.example.com/fleet-uid=' + uid, '-o', 'json'))['items']

    def operator_pod():
        pods = json.loads(k('-n', 'celld-system', 'get', 'pods', '-l', 'app.kubernetes.io/name=celld-operator', '-o', 'json'))['items']
        # The harness also labels a probe pod in this namespace; keep only the Deployment's pods.
        live = [p for p in pods if not p['metadata'].get('deletionTimestamp') and p['metadata']['name'].startswith('celld-operator-')]
        assert len(live) == 1, [p['metadata']['name'] for p in live]
        return live[0]

    def operator_restarts():
        return sum(s.get('restartCount', 0) for s in operator_pod()['status'].get('containerStatuses', []))

    def crashed_at(point):
        pod = operator_pod()['metadata']['name']
        try:
            previous = k('-n', 'celld-system', 'logs', pod, '--previous')
        except RuntimeError:
            return False
        return 'injected crash at lifecycle fault point "%s"' % point in previous

    def hold(seconds, invariant, description):
        until = time.monotonic() + seconds
        while time.monotonic() < until:
            assert invariant(), 'violated during hold: ' + description
            time.sleep(2)
        print('PASS:', description, 'held for', seconds, 's', flush=True)

    def write_readable(pod, fleet):
        return json.loads(curl(pod, fleet, '/?cell=integration&id=ack', 8080))['stored'] is True

    # ---- 1. crash after the replica effect, before the journal transition ----
    wait_for(lambda: settled('alpha', 2), 'alpha at baseline before crash injection', timeout=240)
    generation = get('deployment', 'alpha')['metadata']['generation']
    set_operator_fault('after-effect')
    k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': 3}}))
    wait_for(lambda: get('deployment', 'alpha')['spec']['replicas'] == 3 and operator_restarts() >= 1 and crashed_at('after-effect'),
             'manager exited after the replica CAS with the journal still at the prior count', timeout=240)
    issued = get('deployment', 'alpha')['metadata']['annotations'][OPERATION]
    # The restarted manager reconstructs the issued effect from the workload alone.
    wait_for(lambda: settled('alpha', 3) and journal('alpha')['Applied'] == 3, 'restart reconstructs the issued addition', timeout=240)
    assert len(completions('alpha', issued)) == 1, completions('alpha', issued)
    assert get('deployment', 'alpha')['metadata']['generation'] == generation + 1, 'more than one workload spec write'
    set_operator_fault(None)
    print('PASS: crash after effect: one replica write, one completion record', flush=True)

    # ---- 2. crash after durable intent, before the replica effect ----
    generation = get('deployment', 'alpha')['metadata']['generation']
    set_operator_fault('before-effect')
    k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': 2}}))
    wait_for(lambda: operator_restarts() >= 1 and crashed_at('before-effect'), 'manager exited after journaling contraction intent, before the CAS', timeout=420)
    intent = journal('alpha')['Operation']
    assert intent and intent['From'] == 3 and intent['To'] == 2, intent
    assert get('deployment', 'alpha')['spec']['replicas'] == 3
    assert get('deployment', 'alpha')['metadata']['generation'] == generation, 'effect issued despite the crash'
    hold(20, lambda: get('deployment', 'alpha')['spec']['replicas'] == 3 and get('deployment', 'alpha')['metadata']['generation'] == generation,
         'crash-looping manager never issues the effect')
    set_operator_fault(None)
    wait_for(lambda: settled('alpha', 2), 'recovered manager issues the journaled contraction exactly once', timeout=420)
    assert get('deployment', 'alpha')['metadata']['generation'] == generation + 1, 'more than one workload spec write'
    assert len(completions('alpha', intent['ID'])) == 1, completions('alpha', intent['ID'])
    assert write_readable('client', 'alpha')
    print('PASS: crash before effect: intent retained, one decrement after restart, write readable', flush=True)

    # ---- 3. node loss for a Bucket replica ----
    k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': 3}}))
    wait_for(lambda: settled('alpha', 3), 'alpha at three before node loss', timeout=240)
    victim = next(p for p in fleet_pods('alpha') if p['spec']['nodeName'] == cluster + '-worker2')
    k('cordon', cluster + '-worker2')
    try:
        k('-n', 'fleets', 'delete', 'pod', victim['metadata']['name'], '--wait=false')

        def replacement_pending():
            pods = fleet_pods('alpha')
            pending = [p for p in pods if p['status'].get('phase') == 'Pending' and not p['metadata'].get('deletionTimestamp')]
            return len(pods) >= 3 and any(any(c.get('reason') == 'Unschedulable' for c in p['status'].get('conditions', [])) for p in pending)
        wait_for(replacement_pending, 'strict hostname separation leaves the replacement Pending on a cordoned cluster', timeout=120)
        generation = get('deployment', 'alpha')['metadata']['generation']
        k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': 2}}))
        time.sleep(15)
        hold(45, lambda: get('deployment', 'alpha')['spec']['replicas'] == 3 and get('deployment', 'alpha')['metadata']['generation'] == generation,
             'contraction is not issued while a replica is unavailable (reason: %s)' % ready_reason('alpha'))
        assert ready_reason('alpha') not in ('Ready', 'Provisioning'), ready_reason('alpha')
    finally:
        k('uncordon', cluster + '-worker2')
    wait_for(lambda: settled('alpha', 2), 'contraction proceeds once every replica is observable again', timeout=420)
    assert write_readable('client', 'alpha')
    print('PASS: Bucket node loss: no decrement while unavailable, write readable after recovery', flush=True)

    # ---- 4. node loss for a launcher-managed PersistentFleet replica ----
    probe('beta-client', {'celld.example.com/client-of': 'beta'})
    curl('beta-client', 'beta', '/?cell=integration&id=ack', 8080, method='PUT')
    wait_for(lambda: settled('beta', 2), 'beta at baseline before node loss', timeout=240)
    host = get('pod', 'beta-1')['spec']['nodeName']
    claim_uid = get('pvc', 'data-beta-1')['metadata']['uid']
    sts_generation = get('statefulset', 'beta')['metadata']['generation']
    k('cordon', host)
    try:
        k('-n', 'fleets', 'delete', 'pod', 'beta-1', '--wait=true')
        wait_for(lambda: get('pod', 'beta-1')['status'].get('phase') == 'Pending', 'recreated ordinal waits for its retained host', timeout=120)
        hold(30, lambda: get('statefulset', 'beta')['spec']['replicas'] == 2 and get('pvc', 'data-beta-1')['metadata']['uid'] == claim_uid
             and get('statefulset', 'beta')['metadata']['generation'] == sts_generation,
             'no decrement, template change or claim replacement while the ordinal is unschedulable')
        assert not ready('beta')
    finally:
        k('uncordon', host)
    wait_for(lambda: settled('beta', 2) and get('pod', 'beta-1')['spec']['nodeName'] == host, 'ordinal returns on its original host with the same claim', timeout=300)
    assert get('pvc', 'data-beta-1')['metadata']['uid'] == claim_uid
    assert write_readable('beta-client', 'beta')
    print('PASS: PersistentFleet node loss: retained claim, same host, write readable', flush=True)

    # ---- toxiproxy control ----
    def toxic(method, path, body=None):
        args = ['-n', 'celld-test-store', 'exec', 'toxi-ctl', '--', 'curl', '--fail', '--silent', '--show-error', '--max-time', '5', '-X', method, 'http://toxiproxy-api:8474' + path]
        if body is not None:
            args += ['-H', 'Content-Type: application/json', '-d', json.dumps(body)]
        return k(*args)

    proxies = json.loads(toxic('GET', '/proxies'))
    assert proxies['minio']['enabled'] and proxies['minio']['listen'].endswith(':9000'), proxies

    # ---- 5. S3 latency ----
    toxic('POST', '/proxies/minio/toxics', {'name': 'latency', 'type': 'latency', 'stream': 'downstream', 'attributes': {'latency': 250, 'jitter': 50}})
    try:
        started = time.monotonic()
        k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': 3}}))
        wait_for(lambda: settled('alpha', 3), 'growth under 250 ms S3 latency', timeout=300)
        k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': 2}}))
        wait_for(lambda: settled('alpha', 2), 'contraction under 250 ms S3 latency', timeout=420)
        assert write_readable('client', 'alpha')
        print('PASS: S3 latency: shrink/grow completed in %.0f s, write readable' % (time.monotonic() - started), flush=True)
    finally:
        toxic('DELETE', '/proxies/minio/toxics/latency')

    # ---- 6. S3 partition during an issued contraction ----
    k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': 3}}))
    wait_for(lambda: settled('alpha', 3), 'alpha at three before partition', timeout=300)
    applied_before = journal('alpha')['Applied']
    k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': 2}}))
    deadline = time.monotonic() + 420
    caught = None
    while time.monotonic() < deadline:
        deployment = get('deployment', 'alpha')
        op = journal('alpha').get('Operation')
        if deployment['spec']['replicas'] == 2 and op:
            caught = op
            break
        if not op and deployment['spec']['replicas'] == 2:
            break
        time.sleep(0.5)
    toxic('POST', '/proxies/minio/toxics', {'name': 'partition', 'type': 'timeout', 'stream': 'upstream', 'attributes': {'timeout': 0}})
    try:
        if caught:
            print('Partition injected with operation', caught['ID'], 'in phase', caught['Phase'], flush=True)

            def frozen():
                j = journal('alpha')
                op = j.get('Operation')
                return (get('deployment', 'alpha')['spec']['replicas'] == 2 and j['Applied'] == applied_before and j.get('Loss', '') == ''
                        and op is not None and op['ID'] == caught['ID'] and not completions('alpha', caught['ID']))
            hold(45, frozen, 'issued contraction neither completes, decrements again, nor records loss while S3 is unreachable')
        else:
            print('Partition window missed: contraction completed before injection; testing steady-state outage', flush=True)
            hold(45, lambda: get('deployment', 'alpha')['spec']['replicas'] == 2 and journal('alpha').get('Loss', '') == '',
                 'no replica or loss change while S3 is unreachable')
    finally:
        toxic('DELETE', '/proxies/minio/toxics/partition')
    wait_for(lambda: all(p['status'].get('phase') == 'Running' and any(c['type'] == 'Ready' and c['status'] == 'True' for c in p['status'].get('conditions', []))
                         for p in fleet_pods('alpha') if not p['metadata'].get('deletionTimestamp')),
             'runtime pods serve again after the partition (self-fenced processes restarted)', timeout=300)
    assert write_readable('client', 'alpha')
    # Liveness after a partition is not promised: the operator may complete or
    # stay blocked on replaced generations. Safety is: one effect, no false completion.
    outcome_deadline = time.monotonic() + 240
    while time.monotonic() < outcome_deadline and journal('alpha').get('Operation'):
        time.sleep(3)
    final = journal('alpha')
    assert get('deployment', 'alpha')['spec']['replicas'] == 2
    if caught:
        assert len(completions('alpha', caught['ID'])) <= 1
        if final.get('Operation'):
            print('Post-partition outcome: operation %s remains open, reason %s (documented liveness limit)' % (caught['ID'], ready_reason('alpha')), flush=True)
            assert final['Applied'] == applied_before
        else:
            print('Post-partition outcome: operation completed after evidence returned', flush=True)
            assert final['Applied'] == 2 and len(completions('alpha', caught['ID'])) == 1
    print('PASS: S3 partition: fail closed throughout, acknowledged write readable after recovery', flush=True)
    print(k('get', 'celldstoragereservations', '-o', 'json'), flush=True)
